### DISCUSSION TITLE
Term search feature on site

### DISCUSSION THREAD

**Discussion by AbrarNitk** [fastn-stack/fastn#733] [Ideas & RFCs]
Feature: [terms](https://github.com/FifthTry/fpm/discussions/434)

<details>

<summary>Status: Under discussion</summary>

- [ ] Store ID Map in .build/-/id.json when fpm build is called.
- [ ] Create another PR which serves this term map on /-/id.json URL when we do fpm serve.
- [ ] Then finally build a UI search using Id.json data.

----
</details>

We should implement search.

# Choice 1: `Client Side Search`

In this solution we do not have any dependency on anything on server side. If we do this search on `fpm build` powered site will continue to work with zero external dependency. 

We generate the index during `fpm build`, and pass it to frontend, maybe as a separate JSON file which gets loaded when user hovers over the search button or triggers the search keyboard shortcut. We can use one of [tinysearch](https://endler.dev/2019/tinysearch?ref=hackernoon.com), [stork](https://stork-search.net), [lunr](https://lunrjs.com/docs/index.html) etc for this.

# Choice 2: `fpm powered search`

If a fpm project is hosted via fpm controller, then fpm controller can be used to power search. Here there are a variety of self search indexing solution, and variety of third party search options.

The request to search will land with our fpm controller. We can use elastic search, etc, open source services. Or we can use PostgreSQL full text search. Or Lucene etc. 

We have a single tenant solution, where we have an infra, a virtual machine, dedicated to a fpm package. On that machine we can store our search index, run our search service. Since we want to keep the virtual machine footprint small, it may make sense to not use elastic search, but something single node, maybe less powerful than elastic, so we can have small memory footprint. 

Since each package is small, and since package change happen in a controlled fashion (using fpm sync api), we can create/update index one time, and load it in fpm controller memory, or maybe even fpm memory, maybe entire search is inside fpm, we do not have another service at all for searching. 

# Choice 3: `External Search`

Third option is we use some hosted search, eg algolia. We send our fpm documents, either raw or converted to HTML/json to them, and let them index it, and all client side search requests goes directly from browser to algolia etc. 

# All Three?

It's not really an either or situation. As long as we have `fpm build` as a first class supported thing, as long as we decide to support search for `fpm build`, we need a client search capability. 

Similarly, fpm or fpm-controller implementing search means we do not have dependency on external provider, and for us it may be the cheapest/most scalable solution.

And third part hosted search can be useful for `fpm build` scenario. Also it will be needed if there are search capabilities we can not replicate in fpm search. fpm search, whatever we pick is going to be a single solution, with its own limitation/tradeoffs. If someone really wants the better search provided by algolia or whatever we will need to support that as well.

# Search To Include Non Package Content

There is also a possibility that search is not purely powered by fpm package content, but other data, eg what if you also have a user generated data, or data hosted in another database, and you want your site to have a single/universal search.

We can not index everything with us. But maybe client can be given a capability to feed our content in search index of their choice, eg elastic, and then allow our search UI to be powered by their API.

# Build Time Package Scan

Some fpm feature, eg [term linking feature](https://github.com/FifthTry/fpm/discussions/362), depends on our ability to scan entire package content before we start processing any document. We do something like that for sitemap and assets special modules also. Sitemap contains list of every document referenced from sitemap, and checks if the document exists in the package or not. Assets special module contains information about all non ftd file. 

Can we share work between package scan and search somehow?

# Big Package Support

Currently all packages we have created using fpm are relatively small packages. Can we have a single package for say the English language Wikipedia?

Any feature that depends on package scan has to be re-evaluated if we have a package for millions of documents. 

`client side search` maybe ruled out, but  for 2 and 3, we will need incremental indexing, instead of recreate the index on every page change. Incremental indexing is always desirable, but for small package the complexity of incremental indexing may not be worth it as an optimisation.

**Comment by amitu** [comment]
I was research [nextra, a next powered static site generate by vercel](https://nextra.vercel.app) and they implement search by only indexing the [page title of all the pages](https://github.com/shuding/nextra/commit/7e25f357689314f3568fbe68413e02bbc49c172c#diff-7eef2c4322da6b9a962b2242f41463b5a2db77d7d68fa9592b9b3bf4531493d6R39-R41). They plan to search in all headings, and maybe full text. 

They use [match-sorter](https://www.npmjs.com/package/match-sorter) for searching in an array of strings. They also use [remove-accents](https://www.npmjs.com/package/remove-accents) to normalise strings for before searching.

Funnily they do not mention search in their documentation. The only place search shows up is in a commit. Maybe they do not want to acknowledge that the search is title only.
